rfcs/12-primitive-shapes.md

```diff
@@ -0,0 +1,545 @@
+
+Meshing could be naturally extended with other libraries or parameters. For example, a sphere by default might use `Icosphere`, but could be extended to generate UV spheres or quad spheres. For 2d types, we could use a crate such as `lyon` to generate 2d meshes from parameterized primitive shapes within the `Meshable` interface.
+
+### Bounding
+
+For use in spatial queries, broad phase collision detection, and raycasting, hypothetical `Bounding` and `Bounding2d` traits are provided. These traits are implemented for types that define position in addition to the shape and size defined in the base primitive. The basic functionality of this trait is to check whether one bounding shape is contained within another bounding shape.
+
+### Colliders
+
+Colliders can provide intersection checks against other colliders, as well as check if they are within a bounding volume.
+
+### 3D and 2D
+
+This RFC provides independent 2d and 3d primitives. Recall that the purpose of this is to provide lightweight types, so there are what appear to be duplicates in 2d and 3d, such as `Line` and `Line2d`. Note that the 2d version of a line is smaller than its 3d counterpart because it is only defined in 2d. 3d geometry (or 2d with depth) is assumed to be the default for most cases. The names of the types were chosen with this in mind.
+
+## Implementation strategy
+
+### Helper Types
+
+```rust
+/// Stores an angle in radians, and supplies builder functions to prevent errors (from_radians, from_degrees)
+struct Angle(f32);
+```
+### Traits (REFERENCE ONLY!)
+
+**These traits are provided as reference to illustrate how these primitive shape types might be used. The details of implementation and interface should be determined in a separate RFC, PR, or independent prototypes.**
+
+```rust
+trait Meshable{
+  fn mesh(&self) -> Mesh;
+};
+
+trait Bounding {
+  fn within(&self, other: &impl Bounding) -> bool;
+  fn contains(&self, collider: &impl Collider) -> bool;
+}
+
+trait Bounding2d {
+  fn within(&self, other: &impl Bounding2d) -> bool;
+  fn contains(&self, collider: &impl Collider2d) -> bool;
+}
+
+trait Collider {
+  fn collide(&self, other: &impl Collider) -> Option(Collision);
+  fn within(&self, bounds: &impl Bounding) -> bool;
+}
+
+trait Collider2d {
+  fn collide(&self, other: &impl Collider2d) -> Option(Collision2d);
+  fn within(&self, bounds: &impl Bounding2d) -> bool;
+}
+```
+
+### 3D Geometry Types
+
+```rust
+struct Point(Vec3)
+
+/// Vector direction in 3D space that is guaranteed to be normalized through its getter/setter.
+struct Direction(Vec3)
+impl Meshable for Direction {}
```

Comment:
What does this `Meshable` implementation return? I suppose you could argue it's equivalent to the implementation of `Meshable` for a unit vector, but the mathematician in me wants to point out that storing a `Direction` as a unit vector is just a convention, and a `Direction` is "really" an equivalence class of vectors with no inherent length, not even one. Feel free to ignore this if it's too pedantic, though.

Reply:
No harm in pedantry. 😆 
As I commented above, the goal for that trait was to output a "sane" mesh that is useful for debugging. It doesn't need to be a perfect representation, just enough to help debugging. So, for a `Direction`, this might be a little arrow or a simple line to convey the direction.

Reply:
I see. It seems like in that case the implementations of `Meshable` (for `Direction` and also the three primitives I mentioned in the other comment) would be serving a similar role to `Debug` in the `Debug`/`Display` dichotomy. I wonder if using a similar two-trait system would make sense? Something like, following @cart's suggestion above:
```rust
Sphere: DebugMesh
UvSphere: DebugMesh + DisplayMesh
```
(The actual names of the traits should be bikeshedded.)